Sync Peep Spawn (Updated)

wolfreak99 opened the pull request:
This is a revival of an older PR of mine (#6041), in which I attempted to allow multiplayer syncing of setting the peep spawn. Theres still some pieces here and there to work on, and I've been requested to not use any game commands, which I'm still looking into. For now, if someone can test this to see if it works okay on multiplayer (as my current places router does not have any forwarded ports), I'd appreciate that.

Broxzier commented on src/openrct2/libopenrct2.vcxproj:
```diff
@@ -58,7 +58,36 @@
+    <ClInclude Include="core\String.hpp" />
+    <ClInclude Include="core\StringBuilder.hpp" />
+    <ClInclude Include="core\StringReader.hpp" />
+    <ClInclude Include="core\Util.hpp" />
```
Don't add new files from within the IDE, just create the file and reload the project.

wolfreak99 replied:
I plan to fix this once the code portion itself is fixed up.

Broxzier commented on src/openrct2/actions/GameActionCompat.cpp:
```diff
@@ -237,4 +238,35 @@ extern "C"
+        sint32* edi,
+        sint32* ebp)
+    {
+        Guard::Assert(false, "GAME_COMMAND_PLACE_PEEP_SPAWN DEPRECATED");
```
This entire function seems unnecessary, since the entire define doesn't exist on develop yet.

duncanspumpkin replied:
Yeah the reason why these are here on other game actions is to remind folks to not call it. As this is new there is no need

Broxzier commented on src/openrct2/world/Map.cpp:
```diff
@@ -110,6 +110,7 @@ LocationXY16 gMapSelectionTiles[300];
 static LocationXYZ16 gVirtualFloorLastMinLocation;
 static LocationXYZ16 gVirtualFloorLastMaxLocation;
 rct2_peep_spawn gPeepSpawns[MAX_PEEP_SPAWNS];
+//static sint32 _nextPeepSpawnIndex = 0;
```
Remove this

Broxzier commented on src/openrct2/world/Map.cpp:
```diff
@@ -4894,3 +4895,90 @@ bool map_tile_is_part_of_virtual_floor(sint16 x, sint16 y)
+        *edx & 0xFF,
+        (*ebx >> 8) & 0xFF);
+}
+*/
```
And this

Broxzier commented:
You can test it locally without having ports open, by simply opening two instances of the game, and connecting one to the other.
I've tested it on my computer, and it crashes. When calling `GameActions::Execute` in `place_peep_spawn`, it tries to assign a callback function that is not set.

```
>openrct2.dll!GameAction::SetCallback(std::function<void __cdecl(GameAction const *,GameActionResult const *)> cb) Line 157
 openrct2.dll!Network::EnqueueGameAction(const GameAction * action) Line 1526
 openrct2.dll!network_enqueue_game_action(const GameAction * action) Line 3155
 openrct2.dll!GameActions::Execute(const GameAction * action) Line 170
 openrct2.dll!place_peep_spawn(short x, short y, int z, int direction) Line 246
```

ZehMatt commented on src/openrct2/Game.cpp:
```diff
@@ -1711,4 +1720,5 @@ GAME_COMMAND_POINTER * new_game_command_table[GAME_COMMAND_COUNT] = {
     game_command_balloon_press,
     game_command_modify_tile,
     game_command_edit_scenario_options,
+    game_command_place_peep_spawn,
```
This is not needed if there has been no pre-existing game command.

duncanspumpkin commented on src/openrct2/actions/GameActionCompat.cpp:
```diff
@@ -237,4 +238,35 @@ extern "C"
+    {
+        auto gameAction = PlacePeepSpawnAction(x, y, z, direction);
+        auto result = GameActions::Execute(&gameAction);
+        if (result->Error == GA_ERROR::OK)
```
One of the plans with game actions is that we won't use money32 undefined as an error. Just return true or false.

duncanspumpkin commented on src/openrct2/actions/PlacePeepSpawnAction.hpp:
```diff
@@ -0,0 +1,150 @@
+
+        return std::make_unique<GameActionResult>();
+    }
+};
```
Don't forget the new line or janisozaur will eat you.

wolfreak99 commented:
@Broxzier  >I've tested it on my computer, and it crashes. When calling GameActions::Execute in place_peep_spawn, it tries to assign a callback function that is not set.

I'm not entirely sure how to fix this. It happened to me on multiplayer, too. It's been a few months since I've caught up with the games code updates, along with programming in general, as I just don't have much time anymore. Would you know how to fix this? I'm not sure how to assign a callback.

Broxzier commented:
After some more testing I found that the action factory returns a `nullptr`. To fix that, you need to register the new Action first in `src/openrct2/actions/GameActionRegistration.cpp`.

wolfreak99 commented:
Okay, so I merged down some of the commits and then rebased it to update the latest code. I believe I have fixed alot of the code requests, it was a bit of a struggle trying to see if I addressed all the issues (since I've been busy with everything else, It spanned over a few days, and I'm still trying to get back into the github flow again, so please be patient with this), but I believe I did. 

If the code issues were all addressed, I can try to clean up some of the commit log history a bit more.

Broxzier commented on src/openrct2/Game.h:
```diff
@@ -93,6 +93,7 @@ enum GAME_COMMAND
     GAME_COMMAND_BALLOON_PRESS,
     GAME_COMMAND_MODIFY_TILE,
     GAME_COMMAND_EDIT_SCENARIO_OPTIONS,
+    GAME_COMMAND_PLACE_PEEP_SPAWN,
```
I just saw that the game commands that use game actions have a `// GA` here.

Gymnasiast replied:
Indeed. And since this was never a game command in the first place, it should not need to be specified.

Broxzier replied:
Game actions require a unique ID, and up until now the IDs from the command list could be used. Creating a new array with just game actions would be nicer indeed.

Gymnasiast commented:
> The only issue I found is that it cannot be used while the game is paused, while it could be done before.

Considering it is a build action (of sorts), I think it's appropriate to only allow it in pause mode. (Unless 'build in pause mode' is on, of course.)

wolfreak99 commented:
Should I add the pause check in the Query or in the original button down section?

duncanspumpkin commented on src/openrct2/actions/PlacePeepSpawnAction.hpp:
```diff
@@ -0,0 +1,147 @@
+
+    uint16 GetActionFlags() const override
+    {
+        return GameActionBase::GetActionFlags() | GA_FLAGS::EDITOR_ONLY;
```
`GA_FLAGS::ALLOW_WHILE_PAUSED`

duncanspumpkin commented:
@wolfreak99 see my change that will make it work in pause.

wolfreak99 commented:
I was unable to test the code I have pushed locally, so I need a moment to test it with the appveyor build.

Gymnasiast commented on src/openrct2/network/NetworkAction.cpp:
```diff
@@ -223,7 +223,8 @@ const std::vector<NetworkAction> NetworkActions::Actions =
         {
-            GAME_COMMAND_EDIT_SCENARIO_OPTIONS
+            GAME_COMMAND_EDIT_SCENARIO_OPTIONS,
+            GAME_COMMAND_PLACE_PEEP_SPAWN
```
I think scenario options refer to stuff like the goal, pay-per-ride settings, etc. I would put peep spawn permissions in the same category as placing park entrances.

wolfreak99 replied:
I originally thought about doing it that way, but realized the potential for easier abuse. There are less chances of someone getting scenario options permissions than park properties. it's harder to catch abuse on this thing in specific, too, whereas you can easily see someone renaming a park, changing the park price, or moving the park entrance somewhere.  if you trust someone with having the option to modify the pay-per-ride setting, you'd likely trust someone with the ability to change the peep spawn.

Broxzier replied:
I agree with wolfreak here. Looking at what permissiosn they grant exactly, I think it fits better:`PERMISSION_PARK_PROPERTIES`, while `PERMISSION_EDIT_SCENARIO_OPTIONS`

Broxzier replied:
I agree with wolfreak here. Looking at what permissiosn they grant exactly, I think it fits better: `PERMISSION_PARK_PROPERTIES`
- 
- 

`PERMISSION_EDIT_SCENARIO_OPTIONS`
- Set money
- Forbit marketing
- Guests preferences

Broxzier replied:
I agree with wolfreak here. Looking at what permissions they grant exactly, I think it fits better.
`PERMISSION_PARK_PROPERTIES` allows changing settings that you can change from the park window (buying land rights, changing park name, set entrance fee), nothing guests related.
`PERMISSION_EDIT_SCENARIO_OPTIONS` allows changing settings that already have to do with guests generation (spawn rate, initial money, initial happiness).

I don't expect changing guest entry point to really be a abusive issue. Players that have yet to be trusted by a server owner often have neither of those permissions.

I can imagine it being confusing to know which group will grant this permission however. @Gymnasiast If it's possible to list them under both categories, do you think that's too inconsistent or causes too much confusion? If not, that might be nice to do here.

Gymnasiast replied:
> Perhaps a sandbox permission group may be better for both of these actions instead.

That sounds better.

wolfreak99 replied:
Would this be able to go into a separate pr? Or would someone be able to modify the pr and help add it? Otherwise I'm not sure when I'll be able to get around to doing it. It may have to wait a few days.

Gymnasiast replied:
In that case, put it under PERMISSIONS_PARK_PROPERTIES.

Gymnasiast commented on src/openrct2/world/Map.h:
```diff
@@ -588,4 +588,6 @@ uint8 tile_element_get_ride_index(const rct_tile_element * tileElement);
 void FixLandOwnershipTiles(std::initializer_list<TileCoordsXY> tiles);
 void FixLandOwnershipTilesWithOwnership(std::initializer_list<TileCoordsXY> tiles, uint8 ownership);
 
+bool place_peep_spawn(sint16 x, sint16 y, sint32 z, sint32 direction);
```
Could you make this `bool place_peep_spawn(CoordsXYZD location)`?

Gymnasiast commented on src/openrct2/Game.h:
```diff
@@ -93,6 +93,7 @@ enum GAME_COMMAND
     GAME_COMMAND_BALLOON_PRESS,
     GAME_COMMAND_MODIFY_TILE,
     GAME_COMMAND_EDIT_SCENARIO_OPTIONS,
+    GAME_COMMAND_PLACE_PEEP_SPAWN, // TODO refactor to seperate array for just game actions 
```
`// GA, TODO: refactor to separate array for just game actions`

Gymnasiast commented:
I made the changes I wanted myself. I do think this needs testing once more, in order to verify I haven't messed up something.

Gymnasiast commented:
Appears to work correctly.
